### ISSUE TITLE
Optional plus sign makes writing Jsonnet code error prone

### ISSUE THREAD

**Issue by curusarn** [google/go-jsonnet#496]
Consider following example:
```
❯ cat example.jsonnet    
[
  { key1: "something", key2: "something" }
  { key1: "something_else", key2: "something_else" },
]

❯ jsonnet example.jsonnet
[
   {
      "key1": "something_else",
      "key2": "something_else"
   }
]
```

In the example above, there is a missing comma which results in the two objects being merged together. Since the objects have the same fields, the first object simply disappears from the output. Errors like this one are very hard to spot.

I'm wondering why is the "plus sign" / "merge operator" optional. Requiring the operator would make it obvious if the programmer wants to merge objects or if they just forgot a comma. Is there a recommended way to avoid errors like this one? Am I missing something?

Follow up question:
Would it make sense to have an option for `jsonnetfmt` that produces Jsonnet code that includes the optional "merge operators"?

**Comment by seh** [comment]
Pertinent: #233.

**Comment by sbarzowski** [comment]
> I'm wondering why is the "plus sign" / "merge operator" optional. Requiring the operator would make it obvious if the programmer wants to merge objects or if they just forgot a comma. Is there a recommended way to avoid errors like this one? Am I missing something?

The implicit syntax is quite natural for patterns similar to traditional inheritance:
```
local foo = some_template {
    to_be_overriden: bar;
}
```

That said, this convenience is probably not worth the bugs it has caused in other contexts. That ship has sailed, though. The best we can do is to have linter warnings (#233) or other mechanism to minimize the impact.

> Would it make sense to have an option for `jsonnetfmt` that produces Jsonnet code that includes the optional "merge operators"?

Yes, that would be a good thing. Maybe even better than linter support, because it applies the fixes automatically.

**Comment by PaulRudin** [comment]
Another possibility would be to have a command line option to error on implicit +. That way backwards compatibility works, but if people want to enforce it they can.

**Comment by sbarzowski** [comment]
Generally we want to avoid creating language dialects, as much as possible to keep the community together. Having a switch like that would create a lot of problems e.g. with 3rd party libraries which may or may not be compatible.

What we might do one day is pragma statements which change things locally, only for a specific file. This way we can migrate one file at a time. This still introduces a lot of complexity, because all pragma combinations will need to be supported indefinitely in all interpreter implementations. There is still plenty we can do without doing that, so I wouldn't expect pragmas anytime soon. 

And in the meantime, linter warnings are a good solution, once we add them.

**Comment by seh** [comment]
I'm sorry if I missed this suggestion earlier, but one step we could take is to change _jsonnetfmt_ to add back in an elided plus operator, as a nudge to say, "I know what this means here, but let's make sure that was what you meant."

Do we make guarantees that the output of _jsonnetfmt_ will remain stable over time? Would this be a breaking change?

**Comment by sbarzowski** [comment]
> Do we make guarantees that the output of jsonnetfmt will remain stable over time? Would this be a breaking change?

No, we don't promise that it's stable. We prefer it to be stable-ish for obvious reasons. As long as we don't change semantics of the program, we're fine. And in this particular case, I'll be happy with changing it automatically with `jsonnetfmt`.

**Comment by sbarzowski** [comment]
If someone wants to pick this up, it should be quite straightforward and it's enough to implement it in the Go version of `jsonnetfmt`.

Basically, what needs to be done is replacing `ApplyBrace` (for implicit +) node in the syntax tree with `Binary` node (for binary operators).

We already have autoformatting in the opposite direction, applied only when the left hand side is a plain variable reference (like in `some_template { foo: bar }`). It can be used for inspiration: https://github.com/google/go-jsonnet/blob/v0.17.0/internal/formatter/fix_plus_object.go. 

Here's the function which decides which modifications to apply depending on formatter options: https://github.com/google/go-jsonnet/blob/master/internal/formatter/jsonnetfmt.go#L138.

I'll be happy to provide additional assistance if needed.

**Comment by Duologic** [comment]
https://github.com/google/go-jsonnet/pull/506#issuecomment-765438559
> It might become the default later, but let's not get ahead of ourselves :-).

Is now later? :innocent: 

At Grafana Labs, this implicit plus has been considered a nuisance. I just discovered the `--no-use-implicit-plus` option, however it puts the burden on the maintainer rather than the language to set the standard for an ecosystem. Pushing for this from within Grafana towards our open source libraries feels a bit like flying solo on this. I'd much rather have buy in from upstream to set the standard.

wdyt? @sbarzowski

**Comment by sparkprime** [comment]
Fine with me.

**Comment by sparkprime** [comment]
There are probably a bunch of examples etc that would need updating, so it's not just the flick of a switch. But worthwhile I think.

**Comment by sbarzowski** [comment]
Fine with me as well, The main cost here is people's code suddenly breaking. We should mention the workaround in the error message (if it's not there already).

**Comment by ghostsquad** [comment]
I'm looking forward to this change. Teaching folks Jsonnet, and that "implicit" merge has been more of a gotcha, additional mental overhead, and source of hard to find bugs, rather than being super useful.
